### ISSUE TITLE
Subtitles are included in screenshots even when it's disabled in settings

### ISSUE THREAD

**Issue by thesla13** [iina/iina#4582]
Hello!

The title pretty much sums up the issue. Whether or not the option to include the subtitles in screenshots is checked they are present in the captured images ((

**Comment by low-batt** [comment]
How are you initiating taking a screenshot?
Clicking the `Take a screenshot` menu item in the `Playback` menu?
Pressing a key?
If using a key, in IINA's settings on the `Key Bindings` tab, what configuration are you using?

If you are pressing the "s" key and using the mpv key bindings then I think I reproduced the problem you are reporting.

The feature in question is the `Include subtitles` setting found in the `Screenshots` section of the `General` tab in IINA's settings.

Many of IINA's settings directly control mpv options. This one does not. IINA is trying to add functionality on top of mpv and there are problems with the implementation.

The mpv feature in question is the [screenshot](https://mpv.io/manual/stable/#command-interface-screenshot-%3Cflags%3E) command:
> **screenshot \<flags>**
>    Take a screenshot.
>
>    Multiple flags are available (some can be combined with +):
>
>    **\<subtitles> (default)**
>        Save the video image, in its original resolution, and with subtitles. Some video outputs may still include the OSD in the output under certain circumstances.
>    **\<video>**
        Like subtitles, but typically without OSD or subtitles. The exact behavior depends on the selected video output.
>    **\<window>**
        Save the contents of the mpv window. Typically scaled, with OSD and subtitles. The exact behavior depends on the selected video output.
> **\<each-frame>**
>    Take a screenshot each frame. Issue this command again to stop taking screenshots. Note that you should disable frame-dropping when using this mode - or you might receive duplicate images in cases when a frame was dropped. This flag can be combined with the other flags, e.g. `video+each-frame`.

From the above documentation we can guess that IINA is trying to avoid having multiple menu items (`Screenshot with subtitles` / `Screenshot without subtitles`) and instead have a setting that tells it what argument to pass when sending the `screenshot` command to mpv.

The special IINA processing is found in `PlayerCore`:
https://github.com/iina/iina/blob/1af17dbc795f602c268afccc8599906ca25a6a9f/iina/PlayerCore.swift#L755-L764

One thing I would change is the `guard` statement that returns if neither `saveToFile` nor `saveToClipboard` have been enabled. That statement should emit a log message indicating that is why a screenshot was not taken.

This code also shows a second IINA customization. The `replyUserdata` specified allows IINA to recognize the reply sent by mpv when the asynchronous `screenshot` command finishes. When that happens the `PlayerCore` method `screenshotCallback` is called. One of the features that provides is the prevue of the screenshot in the OSD.

In my testing this worked for me until I switched to the mpv key bindings. Searching for "screenshot" shows the following bindings:

| Key | Action |
| --- | --- |
| s | screenshot |
| S | screenshot video |
| Ctrl+s | screenshot window |
| Alt+s | screenshot each-frame |

The inclusion of `screenshot window` is a defect. That command is tied to the mpv UI. Executing it fails:
```text
[ 534.841][d][cplayer] Run command: screenshot, flags=73, args=[flags="window", legacy="unused"]
[ 534.841][e][cplayer] Taking screenshot failed.
```

In IINA 1.3.2 pressing "s" invokes the menu item `Take a Screenshot`, which calls the `PlayerCore` method. In IINA 1.3.3 pressing "s" does not invoke the menu item even though that item still shows a keyboard shortcut. This is a regression.

Key presses that do not trigger menu items are processed by the `handleKeyBinding` method in `PlayerWindowController`. That code is missing special handling IINA is attempting to layer on top of mpv and sends the command directly to mpv. This also explains why a preview of the screenshot is not shown in the OSD.

The `handleKeyBinding` method needs to be changed to provide IINA's special processing when the key has been bound to the `screenshot` command. The code in `PlayerCore` will need to be updated to handle an argument having already been supplied.

**Comment by thesla13** [comment]
Hey, low-batt!

Thanks so much for the detailed reply. 

Yes, I am using a keyboard shortcut to capture a screenshot and my shortcuts configuration is a slightly modified "iina default".

So this is indeed a bug and you would look into resolving it in future updates?

Thanks so much!

**Comment by low-batt** [comment]
Not only a bug, but a regression, as this used to work. Regressions are considered a high priority to get fixed as can be seen by the red labels attached to this issue.

Your taking the time to report this is very much appreciated.

I'm a newbie junior developer and don't have the final say on what goes into releases, but in general it is considered critical to fix regressions when possible. The "when possible" caveat has to do with regressions in the libraries from other projects that IINA is built on top off. In such cases IINA may be stuck waiting for a fix from another project. But that is not the case for the issue at hand.

The regression may have already been addressed in a pull request from @svobs. I'm tied up with other work and have not had a chance to check that. I will need to coordinate with @svobs on which of us handles the additional problems specific to this issue identified above. IINA needs to be changed so that when I key binding passes a parameter to the mpv `screenshot` command that flag is passed through to mpv.

I did not discuss `screenshot each-frame` because I did not have time to investigate it. It seemed like that was not working as well. That either needs to be fixed, or the key binding should be removed from the mpv defaults.

**Comment by low-batt** [comment]
@svobs Is there a PR that addresses this one?

**Comment by low-batt** [comment]
Seems to be fixed by PR #4567

**Comment by low-batt** [comment]
A partial fix for this issue has been merged into the IINA `develop` branch. GitHub [automatically closed](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) the linked issue in reaction to the merge. 

I am reopening this issue as the remaining problems still need to be fixed. See the discussion in PR #4567.

**Comment by low-batt** [comment]
@svobs The primary problem has now been fixed, but there are still problems with the command variations like `screenshot video` as discussed in the PR. Did you want to handle the remaining problems, or do you want me to try and fix them?

**Comment by low-batt** [comment]
@svobs Did you want to take this one, or do you want me to handle it?

**Comment by svobs** [comment]
@low-batt Seems like lots to fix with the screenshot OSD in general... Pretty busy lately so I want to confirm the scope of what I'm agreeing to.

I can (1) ensure that the following table is correct:

| Command | Shows IINA OSD? | Includes Subtitles? |
| ------------- | ------------- |------------- |
| `screenshot` | Yes | Use `Include subtitles` setting  |
| `screenshot each-frame` | Yes | Use `Include subtitles` setting  |
| `screenshot subtitles` | Yes | Yes |
| `screenshot video` | Yes | No |

And (2) also remove `screenshot window` from default bindings configurations as you noted above.

**Comment by low-batt** [comment]
That table looks correct to me. You will see I entered issue #4822 for the `screenshot window` problem. The plan is we pull the key binding now and revisit the problem later.

I'm expecting the fix involves enhancing `PlayerWindowController.handleKeyBinding` and `PlayerCore.screenshot`. Is that what you are thinking?

**Comment by svobs** [comment]
> That table looks correct to me. You will see I entered issue #4822 for the screenshot window problem. The plan is we pull the key binding now and revisit the problem later.

Sounds good.

> I'm expecting the fix involves enhancing PlayerWindowController.handleKeyBinding and PlayerCore.screenshot. Is that what you are thinking?

Yes, ~~but substitute `PlayerWindowController.handleKeyBinding` with `MenuController.updateKeyEquivalentsFrom` and/or `MenuController.sameKeyAction`.~~

**Comment by low-batt** [comment]
I'm confused. Possibility due to my lack of knowledge of this part of the code. There is no menu equivalent to some of these screenshot variations, so why would it be tied into the menu code?

**Comment by svobs** [comment]
> There is no menu equivalent to some of these screenshot variations, so why would it be tied into the menu code?

You were right in your original comment. I misspoke.

**Comment by low-batt** [comment]
This was fixed in IINA 1.3.5.
